src/python/pants/backend/python/target_types.py

```diff
@@ -86,6 +86,60 @@ def value_or_global_default(self, python_setup: PythonSetup) -> Tuple[str, ...]:
         return python_setup.compatibility_or_constraints(self.value)
 
 
+class UnrecognizedResolveNamesError(Exception):
+    def __init__(
+        self,
+        unrecognized_resolve_names: list[str],
+        all_valid_names: Iterable[str],
+        *,
+        description_of_origin: str,
+    ) -> None:
+        # TODO(#12314): maybe implement "Did you mean?"
+        if len(unrecognized_resolve_names) == 1:
+            unrecognized_str = unrecognized_resolve_names[0]
+            name_description = "name"
+        else:
+            unrecognized_str = str(sorted(unrecognized_resolve_names))
+            name_description = "names"
+        super().__init__(
+            f"Unrecognized resolve {name_description} from {description_of_origin}: "
+            f"{unrecognized_str}\n\nAll valid resolve names: {sorted(all_valid_names)}"
+        )
+
+
+class PythonResolveField(StringField, AsyncFieldMixin):
+    alias = "experimental_resolve"
+    # TODO(#12314): Figure out how to model the default and disabling lockfile, e.g. if we
+    #  hardcode to `default` or let the user set it.
+    help = (
+        "The resolve from `[python-setup].experimental_resolves_to_lockfiles` to use, if any.\n\n"
+        "This field is highly experimental and may change without the normal deprecation policy."
+    )
+
+    def validate(self, python_setup: PythonSetup) -> None:
+        """Check that the resolve name is recognized."""
+        if not self.value:
+            return None
+        if self.value not in python_setup.resolves_to_lockfiles:
+            raise UnrecognizedResolveNamesError(
+                [self.value],
+                python_setup.resolves_to_lockfiles.keys(),
+                description_of_origin=f"the field `resolve` in the target {self.address}",
+            )
+
+    def resolve_and_lockfile(self, python_setup: PythonSetup) -> tuple[str, str] | None:
```

Comment:
This modeling took a lot of iterations. Originally I only had call sites pass the resolve name to `PexFromTargetsRequest`, and it would do the lookup for the corresponding lockfile from `[python-setup]`. This was great for boilerplate, but I think it's crucial the error message for unrecognized resolve names explains the origin of the error. We need to preserve the `Address` of the offending target.

It was less awkward imo to handle this validation here and having the callers pass the resolve_and_lockfile, rather than passing `resolve_and_description_of_origin`